typed: Allow duplicates when we merge

apelisse opened the pull request:
The goal is to ignore duplicates if they are not included in the partial object, and to replace all of the occurences if they are in the partial object.

/assign @liggitt 
/cc @alexzielenski

apelisse commented on typed/merge_test.go:
```diff
@@ -423,6 +467,18 @@ var mergeCases = []mergeTestCase{{
 		`{"atomicList":["a","a"]}`,
 		`{"atomicList":["a","a"]}`,
+	}, {
+		`{"list":[{"key":"a","id":1,"bv":true},{"key":"b","id":2},{"key":"a","id":1,"bv":false,"nv":2}]}`,
+		`{"list":[{"key":"a","id":1,"nv":3}, {"key":"c","id":3},{"key":"b","id":2}]}`,
+		`{"list":[{"key":"a","id":1,"bv":true,"nv":3},{"key":"c","id":3},{"key":"b","id":2}]}`,
```
So this is merging the new item with the first occurence, I'm currently thinking we should replace altogether.

alexzielenski replied:
This case surprised me. From our prior conversations I would not have expected the merge with one of the duplicates. It doesn't seem like as the author of such a resource I'd be very conscious of which of the duplicates I am merging with.


Is it possible to have two instances of the same value but they are owned by different field managers? In that case I almost would expect "MY" version of the duplicate to be merged; but don't think that'd be worth the effort.

apelisse replied:
> Is it possible to have two instances of the same value but they are owned by different field managers?

No.

> This case surprised me. From our prior conversations I would not have expected the merge with one of the duplicates. It doesn't seem like as the author of such a resource I'd be very conscious of which of the duplicates I am merging with.

yeah, I think we just need to replace rather than merge, which means I'll need to know if they are duplicated and replace altogether, I'll update the algorithm.

alexzielenski commented on typed/merge.go:
```diff
@@ -222,29 +228,33 @@ func (w *mergingWalker) visitListItems(t *schema.List, lhs, rhs value.List) (err
 			if _, ok := observedRHS.Get(pe); !ok {
 				// take LHS item
-				lChild, _ := observedLHS.Get(pe)
+				lChild := lhs.AtUsing(w.allocator, lI)
```
This seems equivalent, since we are looking LHS first. Just curious why the change? Is it a noop?

apelisse replied:
Actually, lhs can have multiple times the same PE, so we want to make sure we get the right one (I don't think it's equivalent).

apelisse replied:
Added a comment for that.

alexzielenski commented on typed/merge.go:
```diff
@@ -290,11 +300,9 @@ func (w *mergingWalker) indexListPathElements(t *schema.List, list value.List) (
-			errs = append(errs, errorf("duplicate entries for key %v", pe.String())...)
-			continue
+		if _, found := observed.Get(pe); !found {
+			observed.Insert(pe, child)
```
Is this existence check necessary? Is `observed` a set?

apelisse replied:
observed is a map from pathelement to the actual value.

the value that ends-up being in the PE is the one we're going to use to merge I think, so if we were to override here, I think we would end-up merging with the last item in the list rather than the first. But since we've decided we might not want to do that, I will probably deal with that differently.

alexzielenski commented on typed/merge.go:
```diff
@@ -222,29 +228,33 @@ func (w *mergingWalker) visitListItems(t *schema.List, lhs, rhs value.List) (err
 				lI++
 				continue
+			} else if _, ok := mergedRHS.Get(pe); ok {
+				// we've already merged it with RHS, we don't want to duplicate it, skip it.
```
I'm not sure why this case is necessary. It's hard to track the invariants of `observedRHS` vs `mergedRHS` without going very deep into the code.

alexzielenski commented on typed/merge.go:
```diff
@@ -180,14 +180,18 @@ func (w *mergingWalker) visitListItems(t *schema.List, lhs, rhs value.List) (err
 	out := make([]interface{}, 0, outLen)
 
-	rhsOrder, observedRHS, rhsErrs := w.indexListPathElements(t, rhs)
+	rhsPEs, observedRHS, rhsErrs := w.indexListPathElements(t, rhs)
```
Are the PE's no longer ordered?

apelisse replied:
They are still, but "order" didn't really tell me what they were. I could revert or make it `rhsOrderedPEs` or something.

alexzielenski reviewed (commented):
Except for one tests make a lot of sense and very straightforward.

TBH I could not gain deep sense of the code changes without obtaining degree in SMD. This code could benefit a lot from some more inline comments about the invariants of this algorithm.

liggitt commented on typed/merge.go:
```diff
@@ -180,14 +180,18 @@ func (w *mergingWalker) visitListItems(t *schema.List, lhs, rhs value.List) (err
 	}
```
AI: look for issue reported where someone couldn't update because of managed fields

liggitt commented on typed/merge.go:
```diff
@@ -199,12 +203,14 @@ func (w *mergingWalker) visitListItems(t *schema.List, lhs, rhs value.List) (err
+			if pe.Equals(rhsPEs[rI]) {
 				// merge LHS & RHS items
+				mergedRHS.Insert(pe, struct{}{})
 				lChild, _ := observedLHS.Get(pe)
```
comment that `lChild` will be nil if the LHS had duplicates for pe

liggitt commented on typed/merge.go:
```diff
@@ -222,29 +228,33 @@ func (w *mergingWalker) visitListItems(t *schema.List, lhs, rhs value.List) (err
-			pe := lhsOrder[lI]
+			pe := lhsPEs[lI]
 			if _, ok := observedRHS.Get(pe); !ok {
 				// take LHS item
```
comment that we use lhs.AtUsing instead of observedLHS because observedLHS can contain nil if the lhs had duplicates

liggitt commented on typed/merge.go:
```diff
@@ -222,29 +228,33 @@ func (w *mergingWalker) visitListItems(t *schema.List, lhs, rhs value.List) (err
-			pe := rhsOrder[rI]
+			pe := rhsPEs[rI]
+			mergedRHS.Insert(pe, struct{}{})
 			lChild, _ := observedLHS.Get(pe) // may be nil
```
```suggestion
			lChild, _ := observedLHS.Get(pe) // may be nil if not present or LHS duplicated this item
```

liggitt commented on typed/merge.go:
```diff
@@ -291,10 +301,10 @@ func (w *mergingWalker) indexListPathElements(t *schema.List, list value.List) (
 		if _, found := observed.Get(pe); found {
-			errs = append(errs, errorf("duplicate entries for key %v", pe.String())...)
-			continue
+			observed.Insert(pe, value.NewValueInterface(nil))
```
is it possible to *only* tolerate duplicates making the LHS and keep returning the duplicate error making the RHS?

liggitt reviewed (commented):
tests look good, the change was shockingly small.

mostly requests for clearer comments, one request to keep the duplicate-error-returning bit the same when constructing observedRHS

apelisse commented:
PTAL, I think I've addressed all the commits. Also split the rename of rhsOrder and lhsOrder in a seperate commit to keep the meat of the change isolated.

liggitt commented:
/lgtm

liggitt commented:
Thanks, open the k/k PR to update master and ping me for review.

I suspect we'll want to backport this, not sure if that means picking the fix to other smd branches or not.
